### ISSUE TITLE
segfault with Struct

### ISSUE THREAD

**Issue by schmurfy** [mobiruby/mruby-cfunc#15]
I cannot figure if it is something I did wrong or if this is an issue with cfunc itself, first here is the relevant C headers:

```
  typedef struct {
      unsigned long number;
      unsigned long size;
      sigar_cpu_t *data;
  } sigar_cpu_list_t;

int sigar_cpu_list_get(sigar_t *sigar, sigar_cpu_list_t *cpulist);
```

and here is my test code:

```
lib  = CFunc::call(CFunc::Pointer, "dlopen", "libsigar.dylib", CFunc::Int.new(2))
sigar_cpu_list_get_func = CFunc::call(CFunc::Pointer, "dlsym", lib, "sigar_cpu_list_get")

class CpuListStruct < CFunc::Struct
  define(
      CFunc::UInt64, :number,
      CFunc::UInt64, :size,
      CFunc::Pointer, :data
    )
end

cpus = CpuListStruct2.new
ret = CFunc::call(CFunc::Int, sigar_cpu_list_get_func, sigar, cpus)
```

I get a segfault on the last line, "sigar" is an opaque object which was initialized previously but I know this part works because if I replace the last two lines with the following it works as expected:

```
cpus = CFunc::Pointer.malloc(3 * 8)
ret = CFunc::call(CFunc::Int, sigar_cpu_list_get_func, sigar, cpus)
p [:ret, ret.value]
p CFunc::UInt64.get(cpus)
```

I got this:
[:ret, 0]
4

Which is correct since I have 4 cores.
Any idea why the struct code segfault ? I tried to trace it and here is the result:

``` bash
Reason: KERN_INVALID_ADDRESS at address: 0x0000000000000000
sigar_cpu_list_create (cpulist=0x0) at sigar.c:582
582     cpulist->number = 0;
(gdb) bt
#0  sigar_cpu_list_create (cpulist=0x0) at sigar.c:582
#1  0x000000010020e956 in sigar_cpu_list_get (sigar=0x10013fbe0, cpulist=0x0) at darwin_sigar.c:875
#2  0x00000001000a55b4 in ffi_call_unix64 ()
#3  0x0000000100141130 in ?? ()
```

a null pointer is passed to the c function instead of the structure address :(
I also tried using cpus.addr as the last argument but it crashes later in cfunc_uint64_class_get also trying to access a 0x0 address.

Ps: sorry for the multiple edit, github keyboard shortcuts are stupid...

**Comment by masuidrive** [comment]
if you want to do in C, 

```
sigar_cpu_list_get(&sigar, &cpus)
```

Could you try it?

```
ret = CFunc::call(CFunc::Int, sigar_cpu_list_get_func, sigar.addr, cpus.addr)
```

**Comment by schmurfy** [comment]
I iwill try to check in C later but given the header what you wrote should be right, what I cannot understand is why it does works without the struct. I tried both "sigar" and "sigar.addr" but it does not change anything it works without the struct and crash with it :(

**Comment by masuidrive** [comment]
Could you make small reproduce code and attached the code?

**Comment by schmurfy** [comment]
looks like I managed to make it woks, I have yet to figure out what I did wrong though but I should be fine xD
Sorry for that I will reopen the issue if I get stuck again.

it would be nice to have a minimal documentation in the readme, even some lines of texts just to point users in the right direction. Now I am starting to understand better how cfunc works mostly by looking at the C code below but in some areas I am still clueless as to how things works.
